maturity-level-two/docs/api-security.md

```diff
@@ -0,0 +1,127 @@
+
+Please keep in mind to use validation properties based on things that cannot be changed in a certificate. By this I think about:
+- Certificate Thumprint
+- Certificate Chain
+- Certificate Serial
+- Validity (Expiration time)
+
+Next to that, with this, it is important to have a good framework to update all the certificate you have in your collection as a certificate is only valid for usually a year. It is also a recommendation to not extend this as this might end up in a security risk if the certificate is received by an unwanted party.
+
+The advantage over the previously explained options is that this can give you a certainty over a user, and if the certificates are securely passed from the one party to the other, you have a very secure system.
+
+If you want to implement this in .NET Core, you can speed up things by using the Arcus NuGet package as described [here](https://webAPI.arcus-azure.net/features/security/auth/certificate).
+
+### OAuth2
+OAuth2 has various implementations. From IdentityServer to manage the users/secrets by yourself or Facebook, Google or Microsoft (with Azure AD) where the whole user management is done by the OAuth2 provider itself.
+
+This setup might require you the most effort but is also the most secure (if you choose only one).
+
+The principle is that the user will authenticate to the OAuth2 Authentication server. The server will validate if your credentials are valid credentials for this certain application. If that is the case an `Authorization`-header with a JWT token can be added to the request. The API can then validate if:
+- The user is authentication with the right Authorization Server
+- What the email & basic information of the user is.
+- The roles the user has (based on claims)
+
+And do the appropriate action custom made for that user.
+
+Bearer tokens can only be generated by the authentication server so there is no security risk there.
+Usually a token is only valid for x amount of time and will expire. Every time a user wants to start a new session he needs to generate a new bearer token via the Authentication server.
+
+Usually within the Authentication Server you can easily add and remove users for your application.
+
+## Azure Specifics
+Azure can give you some out of the box features to improve the security of your system. In the list below you can see some examples how this can help.
+
+### API Management
+- Https is by default there on API management
+- [IP-Filtering can be enabled via policy](https://docs.microsoft.com/en-us/azure/API-management/API-management-access-restriction-policies#RestrictCallerIPs)
+- [Client certificate validation can be done via policies](https://docs.microsoft.com/en-us/azure/API-management/API-management-howto-mutual-certificates-for-clients)
+- [API-Keys can be added via a policy](https://docs.microsoft.com/en-us/azure/API-management/API-management-access-restriction-policies#CheckHTTPHeader)
```

Comment:
That's handled on the API-management and for me Azure specific

Reply:
Not sure I get your comment but think I'm talking about consumer API keys and you are talking about upstream API keys, right?

If so, I'd elaborate on that in the listing so that is clear

Reply:
Ok now I get it, why would you use the https://docs.microsoft.com/en-us/azure/API-management/API-management-access-restriction-policies#CheckHTTPHeader and do all the heavy lifting while API Management does this for you?

Reply:
You mean the user / subscription model? - Api key Validation is a totally different thing than that. If not, not sure what you mean

Reply:
Curious how it is different? :)

Reply:
On API-Key validation you only validate on a specific set  of API Keys, unbounded to the users, just a static thing you can return and you can set it different on every operation if you want. On The User/subscription model you have an onboarding process, and the actual API key you cannot choose. So for me that is a big diff :)

Reply:
Now I see your point. Ok I can see the value of manual API key validation but would strongly recommend to use the built-in APIM capability for this rather than re-inventing the wheel.

What do you think @MassimoC?